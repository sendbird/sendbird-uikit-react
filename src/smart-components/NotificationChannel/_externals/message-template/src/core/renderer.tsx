import React from 'react';

import type { BasicProps, ComponentType, ComponentsUnion, GetProperties } from '../types/components';

type FC<Props, ParsedStyle> = (props: BasicProps<Props, ParsedStyle>) => React.ReactElement | null;

type ComponentProperties<Type extends ComponentType> = GetProperties<Type, ComponentsUnion>;

interface RendererOptions<ParsedStyle> {
  views?: {
    [component in ComponentType]?: FC<ComponentProperties<component>, ParsedStyle>;
  };
}

export type Renderer<ParsedStyle> = {
  [component in ComponentType]: FC<ComponentProperties<component>, ParsedStyle>;
};

const FRAGMENT = ({ children }: React.PropsWithChildren) => <React.Fragment>{children}</React.Fragment>;

export function createRenderer<ParsedStyle = object | string>(
  opts?: RendererOptions<ParsedStyle>,
): Renderer<ParsedStyle> {
  return {
    box: opts?.views?.box || FRAGMENT,
    text: opts?.views?.text || FRAGMENT,
    image: opts?.views?.image || FRAGMENT,
    imageButton: opts?.views?.imageButton || FRAGMENT,
    textButton: opts?.views?.textButton || FRAGMENT,
  };
}
