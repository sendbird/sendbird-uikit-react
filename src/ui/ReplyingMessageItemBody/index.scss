@import '../../styles/variables';

.sendbird-replying-message-item-body {
  position: relative;
  display: inline-flex;
  box-sizing: border-box;
  flex-direction: column;
  &.incoming { align-items: flex-start; }
  &.outgoing { align-items: flex-end; }
  max-width: 400px;

  .sendbird-replying-message-item-body__replied-to {
    position: relative;
    display: inline-flex;
    flex-direction: row;
    align-items: center;
    height: 16px;
    padding: 0px 12px;
    .sendbird-replying-message-item-body__replied-to__icon {
      position: relative;
      margin-bottom: 4px;
      margin-right: 4px;
    }
    .sendbird-replying-message-item-body__replied-to__text {
      position: relative;
      max-width: 360px;
      overflow: hidden;
      overflow-x: hidden;
      white-space: nowrap;
      word-break: keep-all;
      text-overflow: ellipsis;
      font-size: 12px;
    }
  }

  .sendbird-replying-message-item-body__replied-message {
    position: relative;
    display: inline-flex;

    // text message
    .sendbird-replying-message-item-body__replied-message__text-message {
      position: relative;
      white-space: pre-wrap;
      word-break: break-all;
      font-size: 12px;
      padding: 8px 12px 16px 12px;
      border-radius: 16px;
      @include themed() {
        background-color: t(bg-1);
      }
      .sendbird-replying-message-item-body__replied-message__text-message__word {
        max-height: 32px;
        overflow: hidden;
        text-overflow: ellipsis;
      }
    }
    // thumbnail message
    .sendbird-replying-message-item-body__replied-message__thumbnail-message {
      position: relative;
      width: 144px;
      height: 108px;
      border-radius: 6.4px;
      overflow: hidden;
      .sendbird-replying-message-item-body__replied-message__thumbnail-message__placeholder {
        position: absolute;
        display: inline-flex;
        width: 100%;
        height: 100%;
        justify-content: center;
        align-items: center;
        .sendbird-replying-message-item-body__replied-message__thumbnail-message__placeholder__icon {
          display: inline-flex;
          justify-content: center;
          align-items: center;
          width: 22px;
          height: 22px;
          border-radius: 50%;
          background-color: var(--sendbird-light-background-50);
        }
      }
      .sendbird-replying-message-item-body__replied-message__thumbnail-message__cover {
        position: absolute;
        display: inline-flex;
        width: 100%;
        height: 100%;
        justify-content: center;
        align-items: center;
        top: 0px;
        background-color: rgba(256, 256, 256, 0.4);
        .sendbird-replying-message-item-body__replied-message__thumbnail-message__cover__icon {
          position: relative;
          display: inline-flex;
          justify-content: center;
          align-items: center;
          width: 22px;
          height: 22px;
          border-radius: 50%;
          background-color: rgba(256, 256, 256, 1);
        }
      }
    }
    // file message
    .sendbird-replying-message-item-body__replied-message__file-message {
      position: relative;
      display: inline-flex;
      flex-direction: row;
      font-size: 12px;
      padding: 8px 12px 16px 12px;
      border-radius: 16px;
      @include themed() {
        background-color: t(bg-1);
      }
      .sendbird-replying-message-item-body__replied-message__file-message__type-icon {
        margin-right: 2px;
      }
      .sendbird-replying-message-item-body__replied-message__file-message__file-name {

      }
    }
  }
}
