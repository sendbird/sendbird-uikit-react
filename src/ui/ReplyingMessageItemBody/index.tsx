import React, { ReactElement, useContext } from 'react';
import { FileMessage, User, UserMessage } from 'sendbird';
import './index.scss';

import Icon, { IconTypes, IconColors } from '../Icon';
import Label, { LabelTypography, LabelColors } from '../Label';
import ImageRenderer from '../ImageRenderer';
import { LocalizationContext } from '../../lib/LocalizationContext';
import { getClassName, getUIKitFileType, getUIKitFileTypes, isSupportedFileView, isVideo, truncateString } from '../../utils';
interface Props {
  message?: UserMessage | FileMessage;
  parentMessageType?: string; // mime type
  parentMessageText?: string;
  parentMessageUrl?: string;
  parentMessageSender: User;
  isByMe?: boolean;
  className?: string | Array<string>;
}

export default function ReplyingMessageItemBody({
  message,
  parentMessageType,
  parentMessageText,
  parentMessageUrl,
  parentMessageSender,
  isByMe = false,
  className,
}: Props): ReactElement {
  const { stringSet } = useContext(LocalizationContext);
  const parentMessageSenderNickname = parentMessageSender?.nickname;
  const currentMessageSenderNickname = message?.sender?.nickname;
  const uikitFileTypes = getUIKitFileTypes();
  const splitUrl = parentMessageUrl.split('/');

  return (
    <div className={getClassName([className, 'sendbird-replying-message-item-body', isByMe ? 'outgoing' : 'incoming'])}>
      <div className="sendbird-replying-message-item-body__replied-to">
        <Icon
          className="sendbird-replying-message-item-body__replied-to__icon"
          type={IconTypes.REPLY}
          fillColor={IconColors.ON_BACKGROUND_3}
          width="12px"
          height="12px"
        />
        <Label
          className="sendbird-replying-message-item-body__replied-to__text"
          type={LabelTypography.CAPTION_2}
          color={LabelColors.ONBACKGROUND_3}
        >
          {`${currentMessageSenderNickname} ${stringSet.REPLIED_TO} ${parentMessageSenderNickname}`}
        </Label>
      </div>
      <div className="sendbird-replying-message-item-body__replied-message">
        {/* text message */}
        {(!parentMessageType && parentMessageText?.length > 0) && (
          <div className="sendbird-replying-message-item-body__replied-message__text-message">
            <Label
              className="sendbird-replying-message-item-body__replied-message__text-message__word"
              type={LabelTypography.BODY_2}
              color={LabelColors.ONBACKGROUND_3}
            >
              {parentMessageText}
            </Label>
          </div>
        )}
        {/* thumbnail message */}
        {(parentMessageType && isSupportedFileView(parentMessageType) && parentMessageUrl) && (
          <div className="sendbird-replying-message-item-body__replied-message__thumbnail-message">
            <ImageRenderer
              className="sendbird-replying-message-item-body__replied-message__thumbnail-message__image"
              url={parentMessageUrl}
              alt={parentMessageType}
              width="144px"
              height="108px"
              placeHolder={(style) => {
                <div className="sendbird-replying-message-item-body__replied-message__thumbnail-message__placeholder" style={style}>
                  <div className="sendbird-replying-message-item-body__replied-message__thumbnail-message__placeholder__icon">
                    <Icon
                      type={isVideo(parentMessageType) ? IconTypes.PLAY : IconTypes.PHOTO}
                      fillColor={IconColors.ON_BACKGROUND_2}
                      width="14px"
                      height="14px"
                    />
                  </div>
                </div>
              }}
            />
            <div className="sendbird-replying-message-item-body__replied-message__thumbnail-message__cover">
              <div className="sendbird-replying-message-item-body__replied-message__thumbnail-message__cover__icon">
                <Icon
                  type={isVideo(parentMessageType) ? IconTypes.PLAY : IconTypes.PHOTO}
                  fillColor={IconColors.ON_BACKGROUND_2}
                  width="14px"
                  height="14px"
                />
              </div>
            </div>
          </div>
        )}
        {/* file message */}
        {((parentMessageType && !isSupportedFileView(parentMessageType) && parentMessageUrl) && (
          <div className="sendbird-replying-message-item-body__replied-message__file-message">
            <Icon
              className="sendbird-replying-message-item-body__replied-message__file-message__type-icon"
              type={{
                [uikitFileTypes.IMAGE]: IconTypes.PHOTO,
                [uikitFileTypes.VIDEO]: IconTypes.PLAY,
                [uikitFileTypes.AUDIO]: IconTypes.FILE_AUDIO,
                [uikitFileTypes.GIF]: IconTypes.GIF,
                [uikitFileTypes.OTHERS]: IconTypes.FILE_DOCUMENT,
              }[getUIKitFileType(parentMessageType)]}
              fillColor={IconColors.ON_BACKGROUND_3}
              width="16px"
              height="16px"
            />
            <Label
              className="sendbird-replying-message-item-body__replied-message__file-message__file-name"
              type={LabelTypography.BODY_2}
              color={LabelColors.ONBACKGROUND_3}
            >
              {truncateString(splitUrl[splitUrl.length - 1])}
            </Label>
          </div>
        ))}
      </div>
    </div>
  );
}
