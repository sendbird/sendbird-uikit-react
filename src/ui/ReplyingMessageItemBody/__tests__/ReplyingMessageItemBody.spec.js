import React from 'react';
// import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';

import ReplyingMessageItemBody from "../index";

describe('ReplyingMessageItemBody', () => {
  it('should do a snapshot test of the ReplyingMessageItemBody DOM', function() {
    const component = renderer.create(
      <ReplyingMessageItemBody
        message={{ sender: { nickname: 'Simon' } }}
        parentMessageType={null}
        parentMessageText="Hello nice to meet you"
        parentMessageUrl={''}
        parentMessageSender={{ nickname: 'Gabie' }}
        isByMe
      />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
