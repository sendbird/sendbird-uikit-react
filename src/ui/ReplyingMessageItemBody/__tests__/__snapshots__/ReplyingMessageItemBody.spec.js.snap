// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReplyingMessageItemBody should do a snapshot test of the ReplyingMessageItemBody DOM 1`] = `
<div
  className=" sendbird-replying-message-item-body outgoing"
>
  <div
    className="sendbird-replying-message-item-body__replied-to"
  >
    <div
      className="sendbird-replying-message-item-body__replied-to__icon sendbird-icon sendbird-icon-reply sendbird-icon-color--on-background-3"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "height": "12px",
          "minHeight": "12px",
          "minWidth": "12px",
          "width": "12px",
        }
      }
      tabIndex="0"
    >
      <test-file-stub />
    </div>
    <span
      className="sendbird-replying-message-item-body__replied-to__text sendbird-label sendbird-label--caption-2 sendbird-label--color-onbackground-3"
    >
      Simon replied to Gabie
    </span>
  </div>
  <div
    className="sendbird-replying-message-item-body__replied-message"
  >
    <div
      className="sendbird-replying-message-item-body__replied-message__text-message"
    >
      <span
        className="sendbird-replying-message-item-body__replied-message__text-message__word sendbird-label sendbird-label--body-2 sendbird-label--color-onbackground-3"
      >
        Hello nice to meet you
      </span>
    </div>
  </div>
</div>
`;
